### ISSUE TITLE
Universe inconsistency if Princeton VST and QuickChick is required

### ISSUE THREAD

**Issue by MSoegtropIMC** [QuickChick/QuickChick#220]
Dear QuickChick Team,

when I write:
```
Require Import QuickChick.QuickChick.
Require Import VST.floyd.proofauto.
```
I get: `Universe inconsistency. Cannot enforce sig.u0 < eq.u0 because eq.u0 <= sig.u0.`

When I write:
```
Require Import VST.floyd.proofauto.
Require Import QuickChick.QuickChick.
```
I get:
```
Universe inconsistency. Cannot enforce eq.u0 = QuickChick.Sets.40 because QuickChick.Sets.40 <= QuickChick.Sets.8
<= Coq.Relations.Relation_Definitions.1 <= sig.u0 < eq.u0.
```

This is with:
```
coq-quickchick           1.5.0       Randomized Property-Based Testing Plugin for Coq
coq-vst                  2.7         Verified Software Toolchain
```
(the versions in Coq Platform 2021.02).

I chose to post this here since the second error message is QuickChick types related while the first is generic, so my helicopter view guess is that this is more likely to be fixable in QuickChick.

@andrew-appel: FYI.

Best regards,

Michael

**Comment by Lysxia** [comment]
The inequality `sig.u0 < eq.u0` seems like the worst offender and it comes from VST.

Still, there may be some way to remove the `sig.u0 = eq.u0` constraint from our side.

**Comment by MSoegtropIMC** [comment]
Yes, your analysis sounds reasonable, I will create a ticket at VST.

**Comment by andrew-appel** [comment]
When I do `Print Universes` on an empty buffer with nothing at all imported, it shows the constraint  `sig.u0 < eq.u0`  .  So I don't understand the claim that this is coming from VST.

**Comment by Lysxia** [comment]
I see something different.

```coq
Print Universes Subgraph (sig.u0 eq.u0).
(*
Prop < Set
Set < sig.u0
    < eq.u0
*)
```

**Comment by andrew-appel** [comment]
I must have misinterpreted the Print Universes output, and in any case I didn't know about `Print Universes Subgraph` (which is new).

So now I have tracked it down to the Qed of `VST.veric.semax_prog.semax_prog_rule`, and I will need to investigate further.

**Comment by andrew-appel** [comment]
The bad constraint (sig.u0 < eq.u0) was introduced by the f_equal tactic. I made a better f_equal tactic that avoids it. See the explanation in commit [07bc0ee](https://github.com/PrincetonUniversity/VST/commit/07bc0ee10eb958ed2dae1f09981148903a5c9b47).

**Comment by andrew-appel** [comment]
Regarding the comment above, "Still, there may be some way to remove the sig.u0 = eq.u0 constraint from [your] side", I would say, go for it!

**Comment by Lysxia** [comment]
Thanks Andrew, that's a nice explanation!

**Comment by anton-trunov** [comment]
Looks like this is a related issue: https://github.com/smtcoq/sniper/issues/4
